src/transaction_builder.js

```diff
@@ -437,65 +405,101 @@ TransactionBuilder.prototype.sign = function (index, keyPair, redeemScript, hash
 
   // no? prepare
   } else {
-    // must be pay-to-scriptHash?
     if (redeemScript) {
-      // if we have a prevOutScript, enforce scriptHash equality to the redeemScript
+      var redeemScriptHash = bcrypto.hash160(redeemScript)
+
+      // if redeemScript exists, it is pay-to-scriptHash
+      // if we have a prevOutScript, enforce hash160(redeemScriptequality)  to the redeemScript
       if (input.prevOutScript) {
         if (input.prevOutType !== 'scripthash') throw new Error('PrevOutScript must be P2SH')
 
-        var scriptHash = bscript.decompile(input.prevOutScript)[1]
-        if (!bufferEquals(scriptHash, bcrypto.hash160(redeemScript))) throw new Error('RedeemScript does not match ' + scriptHash.toString('hex'))
-      }
-
-      var extracted = extractFromOutputScript(redeemScript, kpPubKey)
-      if (!extracted) throw new Error('RedeemScript not supported "' + bscript.toASM(redeemScript) + '"')
+        var prevOutScriptScriptHash = bscript.decompile(input.prevOutScript)[1]
+        if (!bufferEquals(prevOutScriptScriptHash, redeemScriptHash)) throw new Error('Inconsistent hash160(RedeemScript)')
 
-      // if we don't have a prevOutScript, generate a P2SH script
-      if (!input.prevOutScript) {
-        input.prevOutScript = bscript.scriptHashOutput(bcrypto.hash160(redeemScript))
+      // or, we don't have a prevOutScript, so generate a P2SH script
+      } else {
+        input.prevOutScript = bscript.scriptHashOutput(redeemScriptHash)
         input.prevOutType = 'scripthash'
       }
 
-      input.pubKeys = extracted.pubKeys
+      var expanded = expandOutput(redeemScript, kpPubKey)
+      if (!expanded) throw new Error('RedeemScript not supported "' + bscript.toASM(redeemScript) + '"')
+
+      input.pubKeys = expanded.pubKeys
       input.redeemScript = redeemScript
-      input.redeemScriptType = extracted.scriptType
-      input.signatures = extracted.pubKeys.map(function () { return undefined })
+      input.redeemScriptType = expanded.scriptType
+      input.signatures = expanded.signatures
+
+    // no redeemScript
     } else {
       // pay-to-scriptHash is not possible without a redeemScript
       if (input.prevOutType === 'scripthash') throw new Error('PrevOutScript is P2SH, missing redeemScript')
 
-      // if we don't have a scriptType, assume pubKeyHash otherwise
-      if (!input.scriptType) {
```

Comment:
`.scriptType` was some weird half-half of `prevOutType` and `redeemScriptType`,  it was entirely unnecessary.